In [47]:
# Importing necessary packages
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
import datetime as dt

In [48]:
# Reading the data
data = pd.read_excel(r'Online Retail.xlsx')

In [49]:
# Printing the first 5 rows
data.head()

InvoiceNo StockCode                          Description  Quantity  \
0    536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1    536365     71053                  WHITE METAL LANTERN         6   
2    536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3    536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4    536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   

          InvoiceDate  UnitPrice  CustomerID         Country  
0 2010-12-01 08:26:00       2.55     17850.0  United Kingdom  
1 2010-12-01 08:26:00       3.39     17850.0  United Kingdom  
2 2010-12-01 08:26:00       2.75     17850.0  United Kingdom  
3 2010-12-01 08:26:00       3.39     17850.0  United Kingdom  
4 2010-12-01 08:26:00       3.39     17850.0  United Kingdom

In [50]:
# Shape of the data
data.shape

(541909, 8)

In [51]:
# Size of the data
data.size

4335272

In [52]:
# type of data
data.dtypes

InvoiceNo              object
StockCode              object
Description            object
Quantity                int64
InvoiceDate    datetime64[ns]
UnitPrice             float64
CustomerID            float64
Country                object
dtype: object

In [53]:
# To print unique countries in the data
data.Country.unique()

array(['United Kingdom', 'France', 'Australia', 'Netherlands', 'Germany',
       'Norway', 'EIRE', 'Switzerland', 'Spain', 'Poland', 'Portugal',
       'Italy', 'Belgium', 'Lithuania', 'Japan', 'Iceland',
       'Channel Islands', 'Denmark', 'Cyprus', 'Sweden', 'Austria',
       'Israel', 'Finland', 'Bahrain', 'Greece', 'Hong Kong', 'Singapore',
       'Lebanon', 'United Arab Emirates', 'Saudi Arabia',
       'Czech Republic', 'Canada', 'Unspecified', 'Brazil', 'USA',
       'European Community', 'Malta', 'RSA'], dtype=object)

In [54]:
# Count of unique countries 
data.Country.nunique()

38

In [55]:
# Checking for null values
data.isnull().sum()

InvoiceNo           0
StockCode           0
Description      1454
Quantity            0
InvoiceDate         0
UnitPrice           0
CustomerID     135080
Country             0
dtype: int64

#### There are 135080 missing values in CustomerID. Since, the analysis is based on customers.. we are removing those missing values.

In [56]:
# Deleting the rows which contains null values
n_data = data.dropna(axis = 0)

In [57]:
# shape of the data after deleting null values
n_data.shape

(406829, 8)

In [58]:
# Cross checking for null values if any..
n_data.isnull().sum()

InvoiceNo      0
StockCode      0
Description    0
Quantity       0
InvoiceDate    0
UnitPrice      0
CustomerID     0
Country        0
dtype: int64

In [59]:
# printing the count of orders per country
n_data.Country.value_counts().head()

United Kingdom    361878
Germany             9495
France              8491
EIRE                7485
Spain               2533
Name: Country, dtype: int64

In [60]:
(361878/len(n_data))*100

88.95088599878574

#### Most of the customers from United Kingdom which is 89%. 

In [61]:
# Here, I am segmenting only UK customers
uk_data = n_data.loc[n_data['Country'] == 'United Kingdom']

In [62]:
uk_data.head()

InvoiceNo StockCode                          Description  Quantity  \
0    536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1    536365     71053                  WHITE METAL LANTERN         6   
2    536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3    536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4    536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   

          InvoiceDate  UnitPrice  CustomerID         Country  
0 2010-12-01 08:26:00       2.55     17850.0  United Kingdom  
1 2010-12-01 08:26:00       3.39     17850.0  United Kingdom  
2 2010-12-01 08:26:00       2.75     17850.0  United Kingdom  
3 2010-12-01 08:26:00       3.39     17850.0  United Kingdom  
4 2010-12-01 08:26:00       3.39     17850.0  United Kingdom

In [63]:
uk_data.shape, uk_data.size

((361878, 8), 2895024)

In [64]:
uk_data.describe()

Quantity      UnitPrice     CustomerID
count  361878.000000  361878.000000  361878.000000
mean       11.077029       3.256007   15547.871368
std       263.129266      70.654731    1594.402590
min    -80995.000000       0.000000   12346.000000
25%         2.000000       1.250000   14194.000000
50%         4.000000       1.950000   15514.000000
75%        12.000000       3.750000   16931.000000
max     80995.000000   38970.000000   18287.000000

#### There are negative values in the Quantity. So, we have to remove those values.

In [65]:
# Removing the negative values in the Quantity column.
uk_data = uk_data.loc[uk_data['Quantity'] > 0]

In [66]:
# Final shape of the dat after cleaning the data.
uk_data.shape

(354345, 8)

In [67]:
# to print the count of unique values in each column
def unique_counts(uk_data):
    for i in uk_data.columns:
        count = uk_data[i].nunique()
        print(i, ": ", count)
unique_counts(uk_data)

InvoiceNo :  16649
StockCode :  3645
Description :  3844
Quantity :  294
InvoiceDate :  15615
UnitPrice :  403
CustomerID :  3921
Country :  1


In [68]:
# Calculating the total price for each order.
uk_data['Total Price'] = uk_data['Quantity'] * uk_data['UnitPrice']

In [69]:
uk_data.head()

InvoiceNo StockCode                          Description  Quantity  \
0    536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1    536365     71053                  WHITE METAL LANTERN         6   
2    536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3    536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4    536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   

          InvoiceDate  UnitPrice  CustomerID         Country  Total Price  
0 2010-12-01 08:26:00       2.55     17850.0  United Kingdom        15.30  
1 2010-12-01 08:26:00       3.39     17850.0  United Kingdom        20.34  
2 2010-12-01 08:26:00       2.75     17850.0  United Kingdom        22.00  
3 2010-12-01 08:26:00       3.39     17850.0  United Kingdom        20.34  
4 2010-12-01 08:26:00       3.39     17850.0  United Kingdom        20.34

In [70]:
# finding initial and recent invoice dates to find recency
uk_data['InvoiceDate'].min()

Timestamp('2010-12-01 08:26:00')

In [71]:
uk_data['InvoiceDate'].max()

Timestamp('2011-12-09 12:49:00')

In [72]:
# The last invoice date is 2011/12/09. So, we use 10/12/2011 to calculate recency
present_date = dt.datetime(2011,12,10)

In [73]:
# Calculating recency, frequency and monetary values and storing those in a table
rfmTable = uk_data.groupby('CustomerID').agg({'InvoiceDate': lambda x: (present_date - x.max()).days, 'InvoiceNo': lambda x: len(x), 'Total Price': lambda x: x.sum()})
rfmTable['InvoiceDate'] = rfmTable['InvoiceDate'].astype(int)
rfmTable.rename(columns={'InvoiceDate': 'Recency', 
                         'InvoiceNo': 'Frequency', 
                         'Total Price': 'Monetary_value'}, inplace=True)

In [74]:
rfmTable.head()

Recency  Frequency  Monetary_value
CustomerID                                    
12346.0         325          1        77183.60
12747.0           2        103         4196.01
12748.0           0       4596        33719.73
12749.0           3        199         4090.88
12820.0           3         59          942.34

In [75]:
quantiles = rfmTable.quantile(q=[0.5])

In [76]:
segmented_rfm = rfmTable

In [79]:
# providing ranks for RFM values using functions
def RScore(x,p,d):
    if x <= d[p][0.50]:
        return 1
    else:
        return 2
    
def FMScore(x,p,d):
    if x <= d[p][0.50]:
        return 2
    else:
        return 1


In [80]:
# Segmenting and updating the rfm table
segmented_rfm['r_quartile'] = segmented_rfm['Recency'].apply(RScore, args=('Recency',quantiles,))
segmented_rfm['f_quartile'] = segmented_rfm['Frequency'].apply(FMScore, args=('Frequency',quantiles,))
segmented_rfm['m_quartile'] = segmented_rfm['Monetary_value'].apply(FMScore, args=('Monetary_value',quantiles,))

In [81]:
segmented_rfm.head()

Recency  Frequency  Monetary_value  r_quartile  f_quartile  \
CustomerID                                                               
12346.0         325          1        77183.60           2           2   
12747.0           2        103         4196.01           1           1   
12748.0           0       4596        33719.73           1           1   
12749.0           3        199         4090.88           1           1   
12820.0           3         59          942.34           1           1   

            m_quartile  
CustomerID              
12346.0              1  
12747.0              1  
12748.0              1  
12749.0              1  
12820.0              1

In [82]:
# updating RFM score 
segmented_rfm['RFMScore'] = segmented_rfm.r_quartile.map(str) + segmented_rfm.f_quartile.map(str) + segmented_rfm.m_quartile.map(str)

In [83]:
segmented_rfm.head()

Recency  Frequency  Monetary_value  r_quartile  f_quartile  \
CustomerID                                                               
12346.0         325          1        77183.60           2           2   
12747.0           2        103         4196.01           1           1   
12748.0           0       4596        33719.73           1           1   
12749.0           3        199         4090.88           1           1   
12820.0           3         59          942.34           1           1   

            m_quartile RFMScore  
CustomerID                       
12346.0              1      221  
12747.0              1      111  
12748.0              1      111  
12749.0              1      111  
12820.0              1      111

In [84]:
segmented_rfm[segmented_rfm['RFMScore']=='111'].sort_values('Monetary_value', ascending=False).head(10)

Recency  Frequency  Monetary_value  r_quartile  f_quartile  \
CustomerID                                                               
18102.0           0        431       259657.30           1           1   
17450.0           8        337       194550.79           1           1   
17511.0           2        963        91062.38           1           1   
16029.0          38        242        81024.84           1           1   
16684.0           4        277        66653.56           1           1   
14096.0           4       5111        65164.79           1           1   
13694.0           3        568        65039.62           1           1   
15311.0           0       2379        60767.90           1           1   
13089.0           2       1818        58825.83           1           1   
17949.0           1         70        58510.48           1           1   

            m_quartile RFMScore  
CustomerID                       
18102.0              1      111  
17450.0              1      111  
17511.0              1      111  
16029.0              1      111  
16684.0              1      111  
14096.0              1      111  
13694.0              1      111  
15311.0              1      111  
13089.0              1      111  
17949.0              1      111

In [86]:
segmented_rfm[segmented_rfm['RFMScore'] == '111'].describe()

Recency    Frequency  Monetary_value  r_quartile  f_quartile  \
count  1170.000000  1170.000000     1170.000000      1170.0      1170.0   
mean     16.772650   213.000855     4392.477616         1.0         1.0   
std      13.361638   364.607457    12027.578463         0.0         0.0   
min       0.000000    42.000000      652.820000         1.0         1.0   
25%       5.000000    79.000000     1208.835000         1.0         1.0   
50%      14.000000   134.000000     2045.935000         1.0         1.0   
75%      26.000000   237.000000     3787.260000         1.0         1.0   
max      50.000000  7847.000000   259657.300000         1.0         1.0   

       m_quartile  
count      1170.0  
mean          1.0  
std           0.0  
min           1.0  
25%           1.0  
50%           1.0  
75%           1.0  
max           1.0

In [88]:
segmented_rfm['RFMScore'].value_counts()

111    1170
222    1147
122     471
211     429
221     204
112     175
212     168
121     157
Name: RFMScore, dtype: int64

#### The customers are segmented in to eight groups 

In [89]:
(1170/len(segmented_rfm))*100

29.839326702371842

### Pareto 80/20 rule: 80% of your company sales come from 20% of your customers.
#### Nearly 30% of the customers are possessing high Recency, Frequency and monetary values.